### PR TITLE
bug 1260420 - set featured versions by product-defails.m.o

### PR DESCRIPTION
**Pull request by peterbe**
I appreciate this is a bit hard to test manually. But I've done it. Perhaps you can simply trust me with the manual testing.

Here's what I did:
```
$ pg_dump -p 5433 -h localhost -U breakpad_rw -a -t product_versions -t products  breakpad  > /tmp/stagedbstuff.sql
$ psql socorro_integration_test < /tmp/stagedbstuff.sql
```
Now, I have in my local PG a bunch of realistic `product_versions` data. 

Then I run crontabber like this:
```
$ python socorro/cron/crontabber_app.py --admin.conf=../local-crontabber.ini 
```
In my `../local-crontabber.ini` the postgres credentials are set up to point to my local `socorro_integration_test` database. And the last couple of lines of that ini file are these:
```
    jobs='''
    #socorro.cron.jobs.bugzilla.BugzillaCronApp|1d
    socorro.cron.jobs.featured_versions_automatic.FeaturedVersionsAutomaticCronApp|1h
    '''
```

When I run it I get this output:

```
2016-10-18 13:07:06,827 DEBUG - MainThread -  - MainThread - about to run <class 'socorro.cron.jobs.featured_versions_automatic.FeaturedVersionsAutomaticCronApp'>
2016-10-18 13:07:07,514 INFO - MainThread -  - MainThread - Set featured versions for Firefox to: 49.0.1, 50.0b, 51.0a2, 52.0a1
2016-10-18 13:07:08,026 INFO - MainThread -  - MainThread - Set featured versions for FennecAndroid to: 49.0, 50.0b6, 51.0a2, 52.0a1
2016-10-18 13:07:08,544 INFO - MainThread -  - MainThread - Set featured versions for Thunderbird to: 45.4.0, 50.0b1, 51.0a2
2016-10-18 13:07:08,545 DEBUG - MainThread -  - MainThread - successfully ran <class 'socorro.cron.jobs.featured_versions_automatic.FeaturedVersionsAutomaticCronApp'> on 2016-10-18 17:07:06.843628+00:00
2016-10-18 13:07:08,564 INFO - MainThread -  - MainThread - done.
```

The most important thing to note is the featured versions specifically for Firefox are: `49.0.1, 50.0b, 51.0a2, 52.0a1`
That `50.0b` comes from taking `"LATEST_FIREFOX_DEVEL_VERSION"`  from https://product-details.mozilla.org/1.0/firefox_versions.json and carefully drop the number after the `b`.

### CODE REVIEW

**Comment by peterbe** [comment]
ACTUALLY,

Let's be careful about merging this. 
I didn't want to wait any longer so I started coding. 
Let's see what Wayne responds here: https://bugzilla.mozilla.org/show_bug.cgi?id=1260420#c26

If he prefers to NOT set Thunderbird automatically, I'll revise the code.

File: socorro/cron/jobs/featured_versions_automatic.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    def run(self, connection):
+        assert isinstance(self.config.api_endpoint_url, basestring)
+        assert self.config.api_endpoint_url.endswith('/')
+
+        # The @with_single_postgres_transaction decorator makes
+        # sure this cursor is committed or rolled back and cleaned up.
+        cursor = connection.cursor()
+
+        for product in self.config.products:
+            url = '{}{}_versions.json'.format(
```
Comment (adngdb) [inline socorro/cron/jobs/featured_versions_automatic.py]: I don't like that. I think it misses the point of configuration. As I see it, it is quite likely that were the server to change, that file path could change as well. 

How about making the config option the entire URL with a parameter? Something like this: 

```python
required_config.add_option(
    'api_endpoint_url',
    default='https://product-details.mozilla.org/1.0/{product}_versions.json',
)
```

And then here you can do: 

```python
url = self.config.api_url.format(product=product)
```

File: socorro/cron/jobs/featured_versions_automatic.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                )
+            versions = response.json()
+            self._set_featured_versions(
+                cursor,
+                product,
+                versions,
+            )
+
+    def _set_featured_versions(self, cursor, product, versions):
+        # 'product_name' is what's it's called in our product_versions
```
Comment (adngdb) [inline socorro/cron/jobs/featured_versions_automatic.py]: s/what's/what

File: socorro/cron/jobs/featured_versions_automatic.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        featured = set()
+        if product_name == 'Firefox':
+            featured.add(versions['FIREFOX_NIGHTLY'])
+            featured.add(versions['FIREFOX_AURORA'])
+            featured.add(versions['LATEST_FIREFOX_VERSION'])
+            beta = versions['LATEST_FIREFOX_DEVEL_VERSION']
+            # If the beta version is something like '59.0b12'
+            # then convert that to '59.0b' which is basically right-stripping
+            # any numbers after the 'b'.
+            beta = re.sub('b(\d+)$', 'b', beta)
```
Comment (adngdb) [inline socorro/cron/jobs/featured_versions_automatic.py]: That seems quite complex. Why not just ``beta = beta.split('b')[0] + 'b'``?

Comment (adngdb) [inline socorro/cron/jobs/featured_versions_automatic.py]: Yeah, looking at it, maybe your solution is cleaner. Anyway, your call.

File: socorro/cron/jobs/featured_versions_automatic.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    aliases = {}
+    for item in input_string.split(','):
+        if not item.strip():
+            continue
+        key, value = item.split(':')
+        aliases[key.strip()] = value.strip()
+    return aliases
+
+
+class DownloadError(Exception):
```
Comment (peterbe) [inline socorro/cron/jobs/featured_versions_automatic.py]: No, I quite like the idea that they are different. They just happen to have the same name.
Also, the correlations.py code is also dying a slow death. 

Letting this module have its own class make it easy to reason about where the error is coming from.

If this exception does get instansiated, the crontabber app will simply fail, and log the error.

File: socorro/cron/jobs/featured_versions_automatic.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        'api_endpoint_url',
+        default=(
+            'https://product-details.mozilla.org/1.0/'
+        ),
+        doc='URL from which we can download all the featured versions'
+    )
+    required_config.add_option(
+        'products',
+        default='firefox,mobile,thunderbird',
+        from_string_converter=lambda line: tuple(
```
Comment (peterbe) [inline socorro/cron/jobs/featured_versions_automatic.py]: Good catch!

File: socorro/cron/jobs/featured_versions_automatic.py (hunk 1 above)
Comment (peterbe) [inline socorro/cron/jobs/featured_versions_automatic.py]: Much better!

File: socorro/cron/jobs/featured_versions_automatic.py (hunk 3 above)
Comment (peterbe) [inline socorro/cron/jobs/featured_versions_automatic.py]: The regex makes sure the `b\d+` really is the end of the string. Can't do that with the split. No biggie.

**Comment by peterbe** [comment]
I think all review feedback has now been addressed.

File: socorro/cron/jobs/featured_versions_automatic.py [hunk 6]
```diff
@@ -11,6 +12,15 @@
...
 def alias_list_to_dict(input_string):
+    """Return a dict by splitting the input string by ','
+    and splitting each item by a ':' to make it key and value.
+
+    For example::
+
+        >>> alias_list_to_dict('foo: bar, other:thing, ')
+        {'foo': 'bar', 'other': 'thing'}
+
+    """
```
Comment (adngdb) [inline socorro/cron/jobs/featured_versions_automatic.py]: :sunny:

**Pull request merged**